### PR TITLE
Add autest to make sure H2 grace shutdown is working as intended

### PR DESCRIPTION
**Pull request by duke8253**
This PR addresses a few issues:

~- If a `HTTP2_SESSION_EVENT_FINI` is received after a `HTTP2_SESSION_EVENT_SHUTDOWN_INIT`, the `GOAWAY` frame with the correct `last_stream_id` isn't sent out.~
- ~Correctly set and remove the `Connection: close` header for HTTP/2 (this header is only used as a internal flag to indicate a grace shutdown is needed).~
- ~Initiate a grace shutdown when status is 429.~
- Add test to make sure grace shutdown is working as intended.

### CODE REVIEW

File: src/proxy/http2/Http2Stream.cc [hunk 1]
```diff
@@ -802,16 +802,13 @@ Http2Stream::update_write_request(bool call_update)
...
-        const char *value = field->value_get(&len);
-        if (memcmp(HTTP_VALUE_CLOSE, value, HTTP_LEN_CLOSE) == 0) {
-          SCOPED_MUTEX_LOCK(lock, _proxy_ssn->mutex, this_ethread());
-          if (connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NONE) {
-            connection_state.set_shutdown_state(HTTP2_SHUTDOWN_NOT_INITIATED, Http2ErrorCode::HTTP2_ERROR_NO_ERROR);
-          }
+      // Schedule session shutdown if response header has "Connection: close",
+      // then remove "Connection: close" which is a protocol violation for HTTP/2.
+      if (!this->_send_header.is_keep_alive_set()) {
+        this->_send_header.field_delete(MIME_FIELD_CONNECTION, MIME_LEN_CONNECTION);
```
Comment (maskit) [inline src/proxy/http2/Http2Stream.cc:808]: Do we really need to delete the header here? I think it (and other headers that are invalid for H2) will be deleted by `VersionConverter` later.

File: src/proxy/http2/Http2ClientSession.cc [hunk 2]
```diff
@@ -191,16 +191,22 @@ Http2ClientSession::main_event_handler(int event, void *edata)
     break;
 
   case VC_EVENT_ACTIVE_TIMEOUT:
-  case VC_EVENT_INACTIVITY_TIMEOUT:
+  case VC_EVENT_INACTIVITY_TIMEOUT: {
+    if (this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NONE) {
+      this->connection_state.set_shutdown_state(HTTP2_SHUTDOWN_INITIATED, Http2ErrorCode::HTTP2_ERROR_NO_ERROR);
```
Comment (maskit) [inline src/proxy/http2/Http2ClientSession.cc:196]: I may not understand the issue, but the state should not be changed directly here. It should be set by HTTP2_SESSION_EVENT_SHUTDOWN_INIT event handler. Changing the state at multiple places is going to be super confusing.

File: src/proxy/http2/Http2Stream.cc (hunk 1 above)
Comment (duke8253) [reply src/proxy/http2/Http2Stream.cc:808]: I think it does get deleted, added that mostly for testing purpose. I'll remove that.

**Comment by duke8253** [comment]
> > If a HTTP2_SESSION_EVENT_FINI is received after a HTTP2_SESSION_EVENT_SHUTDOWN_INIT, the GOAWAY frame with the correct last_stream_id isn't sent out.
> > If a HTTP/2 session timed out, no grace shutdown is triggered.
> 
> Aren't these immediate close? Why do we want to initiate grace shutdown on timeout?

We got a complaint from a property saying they don't have a way to determine whether a connection is closed, since the HTTP/2 doesn't have the `Connection: close` header, sending a `GOAWAY` when that happens should give them what they want. 

About that timeout part, was doing some testing and wasn't sure whether we need that, left it in there to get some comments.

**Comment by maskit** [comment]
Ok, I wonder why the property can't check if the connection is closed on TCP layer, but let's leave it aside.

The RFC says:

> A server that is attempting to gracefully shut down a connection SHOULD send an initial GOAWAY frame with the last stream identifier set to 231-1 and a [NO_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#NO_ERROR) code. This signals to the client that a shutdown is imminent and that initiating further requests is prohibited. After allowing time for any in-flight stream creation (at least one round-trip time), the server MAY send another GOAWAY frame with an updated last stream identifier. This ensures that a connection can be cleanly shut down without losing requests.

and

> If a connection terminates without a GOAWAY frame, the last stream identifier is effectively the highest possible stream identifier.

So, a client cannot expect a GOAWAY. A connection close without GOAWAY can happen. I don't think current behavior of ATS is wrong.

That said,  the RFC also says:

> An endpoint MAY send multiple GOAWAY frames if circumstances change. For instance, an endpoint that sends GOAWAY with [NO_ERROR](https://datatracker.ietf.org/doc/html/rfc9113#NO_ERROR) during graceful shutdown could subsequently encounter a condition that requires immediate termination of the connection. The last stream identifier from the last GOAWAY frame received indicates which streams could have been acted upon.

I'd take this as "A server can cancel graceful shutdown and send a GOAWAY without waiting for the 1 RTT of grace time, to close a connection immediately". This may be just a different way of saying what you do on this PR, but I think this is what we should do, and I'd take a different approach.

In most cases, `send_goaway_frame()` is called right before we schedule `HTTP2_SESSION_EVENT_FINI` event. So if ATS receives `HTTP2_SESSION_EVENT_FINI`, GOAWAY should be sent. I don't think we need to forcibly progress the shutdown process (we can simply cancel it) to send the second GOAWAY. The only exception is `Http2ClientSession::do_io_close`. It schedules the event without calling `send_goaway_frame()`. But if the event is scheduled in `do_io_close`, it's probably too late to send something anyway.

TLDR; Isn't canceling graceful shutdown process enough?

File: src/proxy/http2/Http2ClientSession.cc (hunk 2 above)
Comment (shinrich) [reply src/proxy/http2/Http2ClientSession.cc:196]: Will sending this with the SHUTDOWN_CONT event below trigger SHUTDOWN_INIT?  Or are you suggesting that this case should send the SHUTDOWN_INIT event instead and rely on that flow to set the appropriate state?

File: src/proxy/http2/Http2ConnectionState.cc [hunk 3]
```diff
@@ -1434,6 +1451,14 @@ Http2ConnectionState::main_event_handler(int event, void *edata)
 
     ink_assert(this->fini_received == false);
     this->fini_received = true;
+
+    if (shutdown_state == HTTP2_SHUTDOWN_INITIATED) {
+      if (shutdown_cont_event) {
+        shutdown_cont_event->cancel();
+      }
+      this->_shutdown_cont();
+    }
```
Comment (shinrich) [inline src/proxy/http2/Http2ConnectionState.cc:1460]: So cleaning up the in progress shutdown events, since we just shutdown cleanly here?

File: src/proxy/http2/Http2ClientSession.cc (hunk 2 above)
Comment (maskit) [reply src/proxy/http2/Http2ClientSession.cc:196]: I was suggesting scheduling SHUTDOWN_INIT event instead. But if canceling grace shutdown works we don't need it.

File: src/proxy/http2/Http2ConnectionState.cc (hunk 3 above)
Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc:1464]: Yes, since we already started a grace shutdown, I think we should finish the whole process.

**Comment by duke8253** [comment]
@maskit forgot to mention one thing, when doing a `curl` command, if no `GOAWAY` frame is sent, it reports `connection left intact` even when ATS says the H2 session is destroyed. I'm not sure that is the correct behavior, or should we even care about that?

**Comment by maskit** [comment]
Closing a connection without sending GOAWAY is not polite, but I don't think it's a violation. Depends on why the connection was closed.

> Endpoints SHOULD always send a GOAWAY frame before closing a connection so that the remote peer can know whether a stream has been partially processed or not. For example, if an HTTP client sends a POST at the same time that a server closes a connection, the client cannot know if the server started to process that POST request if the server does not send a GOAWAY frame to indicate what streams it might have acted on.

> An endpoint might choose to close a connection without sending a GOAWAY for misbehaving peers.

File: include/proxy/http2/Http2Stream.h [hunk 4]
```diff
@@ -80,6 +80,7 @@ class Http2Stream : public ProxyTransaction
   void set_expect_send_trailer() override;
   bool expect_receive_trailer() const override;
   void set_expect_receive_trailer() override;
+  void set_close_connection(HTTPHdr &hdr) const override;
```
Comment (maskit) [inline include/proxy/http2/Http2Stream.h:83]: Unused

Comment (maskit) [reply include/proxy/http2/Http2Stream.h:83]: Ah, nvm. I just realized it's override.

Comment (maskit) [reply include/proxy/http2/Http2Stream.h:83]: Does this relay Connection header value between a request and a response? I think that should not happen.

```
Client <-- (H2) -- ATS <-- (H1) -- Origin
```

When an origin server returns `Connection: close`, ATS should not close the connection between the client and ATS, right? It's just the origin wants to close the connection between ATS and the origin. The client should be allowed to make another request, which can be served from the cache, on the same connection.

Comment (duke8253) [reply include/proxy/http2/Http2Stream.h:83]: This sets the `Connection: close`, but it doesn't decide whether the connection will be closed. `HttpTransact::handle_response_keep_alive_headers` actually decides whether that close will remain in the response or be removed. But before we never set it in the first place, so all the logic there does nothing for h2 connections.

Comment (maskit) [reply include/proxy/http2/Http2Stream.h:83]: Ok, I think I was confused by this line. It seemed like to modify a client response based on a server response, but it doesn't.
https://github.com/apache/trafficserver/blob/cbe91698e2054ddb0ac33d34a92e270fa9194909/src/proxy/http/HttpTransact.cc#L6978

Http1Transaction and Http2Stream do the same thing in set_close_connection. It's not conditional at all. I assume Http3Transaction should do the same. Now I wonder why we don't modify the `heads` directly in HttpTransact. What's the reason for letting transactions to set the header? I think we should remove this confusing function.

Comment (duke8253) [reply include/proxy/http2/Http2Stream.h:83]: This was added in #8178 , maybe at the time we had some issues with our H2 streams and this header was causing trouble?

Comment (maskit) [reply include/proxy/http2/Http2Stream.h:83]: So we are going back to the original behavior (do the same thing for H1 and H2). Are you sure you don't have the issue that was resolved by #8178, with this change?

If we don't have to to differentiate the behavior, I prefer the original code that modifies `heads` directly in HttpTransact.

Comment (shinrich) [reply include/proxy/http2/Http2Stream.h:83]: The intent with #8178 was to not drain the session if the HttpSM logic sets the connection close due to detected transaction failures.  Shutting down the connection in HTTP/1 makes sense since the TCP connection may be polluted in the case of bad request bodies. However, this is not an issue with HTTP/2. An error with a request body should not require shutting down the session due to cross-contamination.

Explicitly setting the Connection: close header in the HTTP/2 processing logic when w want to close the session seems reasonable if it makes our client's lives easier. But blindly interpreting Connection: Close passed from the state machine processing will cause HTTP/2 sessions to shut down needlessly in some cases.

File: src/proxy/http2/Http2ConnectionState.cc [hunk 5]
```diff
@@ -1530,17 +1555,7 @@ Http2ConnectionState::main_event_handler(int event, void *edata)
...
-    // ..., the server can send another GOAWAY frame with an updated last stream identifier
-    if (shutdown_reason == Http2ErrorCode::HTTP2_ERROR_MAX) {
-      shutdown_reason = Http2ErrorCode::HTTP2_ERROR_NO_ERROR;
-    }
-    send_goaway_frame(latest_streamid_in, shutdown_reason);
-    // Stop creating new streams
-    SCOPED_MUTEX_LOCK(lock, this->session->get_mutex(), this_ethread());
-    this->session->set_half_close_local_flag(true);
+    this->_shutdown_cont();
   } break;
```
Comment (maskit) [inline src/proxy/http2/Http2ConnectionState.cc:1559]: Looks like we no longer need a scope here. Let's remove the braces.

**Comment by maskit** [comment]
@duke8253 What's the status of this PR? Is it ready for review?

**Comment by JosiahWI** [comment]
The AuTest dns_host_down failed:
```
file /tmp/sandbox/dns_host_down/ts/log/error.log : DNS lookup should fail - Failed
        Reason: Contents of /tmp/sandbox/dns_host_down/ts/log/error.log did not contains expression: "DNS Error: no valid server http://resolve.this.com/"
        ```

**Comment by JosiahWI** [comment]
[approve ci rocky autest]

**Comment by duke8253** [comment]
@maskit It's ready for another review.

File: src/proxy/http2/Http2ConnectionState.cc (hunk 3 above)
Comment (maskit) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: I still don't get why we need to call `finish_grace_shutdown()` here.
When we schdule `HTTP2_SESSION_EVENT_FINI`, we send `GOAWAY` with `latest_streamid_in`, right? Doesn't this send an extra `GOAWAY`?

- GOAWAY with the big number for graceful shutdown
- GOAWAY with latest_streamid_in at the place we schedule `HTTP2_SESSION_EVENT_FINI`
- GOAWAY with latest_streamid_in here in the `HTTP2_SESSION_EVENT_FINI` handler .

Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: Because when we get a `HTTP2_SESSION_EVENT_FINI` we set the handler to `Http2ConnectionState::state_closed`, which just does `shutdown_cont_event = nullptr`, so the GOAWAY with `latest_streamid_in` never gets sent out.

Comment (maskit) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: We only want to send two GOAWAY frames at most, A) one with the big number, and B) one with `latest_streamid_in`, right?

B is sent when we schedule `HTTP2_SESSION_EVENT_FINI` event. For example:
https://github.com/apache/trafficserver/blob/c49e60f769db8bdd2b662a2fa1fd2f90108a47dd/src/proxy/http2/Http2ConnectionState.cc#L1451-L1455

So when we reach this line, which is in `HTTP2_SESSION_EVENT_FINI` handler, B is already sent (or it's in the send buffer at least). Calling `finish_grace_shutdown` would send third GOAWAY.

Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: yes, but that's just one case, but not for all of them. That's why I check for the `shutdown_state == HTTP2_SHUTDOWN_INITIATED` to make sure that we only send a second GOAWAY with `latest_streamid_in` when we initiated a grace shutdown.

Comment (maskit) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: The only two places where we don't send GOAWAY prior to scheduling `HTTP2_SESSION_EVENT_FINI` are `Http2ClientSession::do_io_close` and `Http2ServerSession::do_io_close`. And those places seem to be too late to send frames anyway because `connection_state.is_state_closed()` is true if `HTTP2_SESSION_EVENT_FINI` is scheduled there.

Even if I miss a place that schedules `HTTP2_SESSION_EVENT_FINI` without sending GOAWAY, or even if we add new code like that, checking `shutdown_state == HTTP2_SHUTDOWN_INITIATED` doesn't make sense to me. How does it prevent sending a third `GOAWAY`? It's unrelated to whether a GOWAY with `latest_streamid_in` is sent. Are you saying the change ensures that there will be at least one GOAWAY frame with `latest_streamid_in` and there may be two of them in some cases, and that's fine?

I hate to add flags just to correct behavior forcibly, but it would be more straightforward solution if we add `bool is_the_last_goaway_sent` and set `true` in `Http2ConnectionState::send_goaway_frame` if the stream id is the big number. Then we can use it to ensure that ATS always send just one GOAWAY frame with `latest_streamid_in`.

Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: The only way we get an event `HTTP2_SESSION_EVENT_SHUTDOWN_INIT` is when `shutdown_state == HTTP2_SHUTDOWN_NOT_INITIATED`, and that is the trigger for a grace shutdown since we only schedule that event when checking for the state like this. 
```
if (this->connection_state.get_shutdown_state() == HTTP2_SHUTDOWN_NOT_INITIATED) {
  send_connection_event(&this->connection_state, HTTP2_SESSION_EVENT_SHUTDOWN_INIT, this);
}
```

And from testing I've done so far, I was never able to get the second GOAWAY frame as described in the RFC when a grace shutdown is initiated, and that's why I created this PR. But I can move `this->_finish_grace_shutdown();` two lines above where it is at now, since `shutdown_cont_event` is only scheduled during a grace shutdown.

Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: Maybe some other changes in the past few months have fixed this in a different way then, I just checked with Proxy Verifier and it is also getting the second GOAWAY frame now. So it's your call then, do you want to add a flag to make sure the second GOAWAY gets sent during a grace shutdown, or I can remove this part of the change and just add the test to check things are working as intended.

Comment (maskit) [reply src/proxy/http2/Http2ConnectionState.cc:1523]: Okay, nice. Then the current code works as I expect. I would not add the flag, because I think something is wrong if the flag has to work.

Let's just add the test. Thank you for writing the test.

Comment (duke8253) [reply src/proxy/http2/Http2ConnectionState.cc]: Sounds good.

File: tests/gold_tests/h2/http2_close_connection.test.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
+'''
+Abort HTTP/2 connection using RST_STREAM frame.
```
Comment (maskit) [inline tests/gold_tests/h2/http2_close_connection.test.py:2]: Please update the comment and the summary (it looks truncated).

**Review by maskit — APPROVED** [review]

**Comment by cmcfarlen** [comment]
Cherry-picked to v10.0.x to avoid conflicts with another PR
